base-server/index.js

```diff
@@ -64,6 +65,48 @@ function normalizeTypeCallbacks(name, callbacks) {
   }
 }
 
+function queueWorker(arg, cb) {
+  let { action, meta, queueKey, server } = arg
+  let queue = server.queues.get(queueKey)
+
+  let end = (...args) => {
+    server.actionsInQueue.delete(meta.id)
+    if (!queue.length()) {
+      server.queues.delete(queueKey)
+    }
+    cb(...args)
+  }
+
+  let undoRemainingTasks = () => {
+    for (let task of queue.getQueue()) {
+      server.undo(task.action, task.meta, 'error')
+    }
+    queue.killAndDrain()
+  }
+
+  let unbindError = server.on('error', (e, errorAction) => {
```

Comment:
It is not very effective to bind new listener on every action (you could have 500 of them in a second).